File: fr.unice.i3s.sigma.delegates/src/fr/unice/i3s/sigma/delegates/SigmaDelegateDomain.java
package fr.unice.i3s.sigma.delegates;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import org.apache.commons.lang3.reflect.TypeUtils;
import org.eclipse.emf.common.util.DelegatingEList;
import org.eclipse.emf.common.util.EList;
import org.eclipse.emf.ecore.EClassifier;
import org.eclipse.emf.ecore.EObject;
import org.eclipse.emf.ecore.EOperation;
import org.eclipse.emf.ecore.EStructuralFeature;
import org.eclipse.emf.ecore.ETypedElement;
import org.eclipse.emf.ecore.EValidator;
import com.google.common.collect.ImmutableList;

import fr.unice.i3s.sigma.core.SigmaConstants;
import fr.unice.i3s.sigma.core.Utils;
import fr.unice.i3s.sigma.core.ValidationResult;

public class SigmaDelegateDomain implements SigmaConstants {

	private static final String SIGMA_DELEGATE_URI = "http://www.i3s.unice.fr/Sigma";

	/**
	 * Registers globally Java version of the Sigma delegates for the
	 * {@link SigmaConstants#SIGMA_DELEGATE_URI} URI. This includes:
	 * <ul>
	 * <li>Invocation delegate {@link SigmaInvocationDelegate}
	 * <li>Setting delegate {@link SettingDelegate}
	 * <li>Validation delegate {@link ValidationDelegate}
	 * </ul>
	 * 
	 * In order to fully leverage the extended return type of the Sigma
	 * validation delegate a {@link SigmaEObjectValidation} must be registered
	 * as the package validator. Normally, this is not the case and it has to be
	 * done explicitly by changing the generate {@code <PackageName>Validator}
	 * super class to be {@link SigmaEObjectValidator}.
	 */
	public static void installGlobally() {
		EOperation.Internal.InvocationDelegate.Factory.Registry.INSTANCE.put(
				SIGMA_DELEGATE_URI, instance.getInvocationDelegateFactory());

		EStructuralFeature.Internal.SettingDelegate.Factory.Registry.INSTANCE
				.put(SIGMA_DELEGATE_URI, instance.getSettingDelegateFactory());

		EValidator.ValidationDelegate.Registry.INSTANCE.put(SIGMA_DELEGATE_URI,
				instance.getValidationDelegateFactory());

	}

	private static final SigmaDelegateDomain instance = new SigmaDelegateDomain();

	public static SigmaDelegateDomain getDefaultInstance() {
		return instance;
	}

	public static <E> EList<E> delegatingEList(final List<E> underlyingList) {
		return new DelegatingEList<E>() {
			private static final long serialVersionUID = -8004643740120935908L;

			@Override
			protected List<E> delegateList() {
				return underlyingList;
			}
		};
	}

	private SigmaValidationDelegateFactory validationFactory;
	private SigmaInvocationDelegateFactory invocationFactory;
	private SigmaSettingDelegateFactory settingFactory;

	public String getURI() {
		return SIGMA_DELEGATE_URI;
	}

	protected String classifierName(EClassifier classifier) {
		String name = classifier.getName();
		if (name == null) {
			name = classifier.getInstanceTypeName();
		}
		return name;
	}

	protected String classifierTypeName(EClassifier classifier) {
		// FIXME: support generics
		return classifierName(classifier);
	}

	protected String elementTypeName(ETypedElement element) {
		String clazzName = element.getEType().getName();
		if (element.isMany()) {
			return Collection.class.getCanonicalName() + "<" + clazzName + ">";
		} else {
			return clazzName;
		}
	}

	protected boolean checkClassifierType(EClassifier eType, Type type) {
		// FIXME: this only work for simple cases without bounds and arrays
		return TypeUtils.isAssignable(type, eType.getInstanceClass());
	}

	protected boolean checkElementType(ETypedElement eType, Type type) {
		if (eType.isMany()) {
			// FIXME: this only work for simple cases without bounds and arrays
			if (type instanceof ParameterizedType) {
				ParameterizedType ptype = (ParameterizedType) type;
				return getCollectionType().isAssignableFrom(
						(Class<?>) ptype.getRawType())
						&& checkClassifierType(eType.getEType(),
								ptype.getActualTypeArguments()[0]);
			} else {
				return false;
			}
		} else {
			// FIXME: Option<T> and others like that are not supported
			return checkClassifierType(eType.getEType(), type);
		}
	}

	protected Class<?> getCollectionType() {
		return Collection.class;
	}

	protected Object processResult(ETypedElement target, Object value) {

		// if the element is multi-reference / multi-attribute it has to be
		// converted to EList
		if (target.isMany() && !(value instanceof EList<?>)) {

			if (!(value instanceof List<?>) && value instanceof Collection<?>) {
				value = ImmutableList.copyOf((Collection<?>) value);
			}

			return delegatingEList((List<?>) value);
		} else {
			return value;
		}
	}

	protected Class<?> loadDelegateClass(String delegateClassName) {
		try {
			return Class.forName(delegateClassName);
		} catch (ClassNotFoundException e) {
			return null;
		}
	}

	public Object processArgument(ETypedElement element, Object value) {
		return value;
	}

	public SigmaDelegateException handleIllegalArgumentException(
			Method delegate, Object object, IllegalArgumentException e) {
		e.printStackTrace();

		// FIXME: if the problem is in implementation of the toString() method
		// this will keep failing
		return new SigmaDelegateException("Unable to invoke delegate method: '"
				+ delegate.toString() + "' on '" + object + "'", e);
	}

	public SigmaDelegateException handleIllegalAccessException(Method delegate,
			Object object, IllegalAccessException e) {
		e.printStackTrace();

		// FIXME: if the problem is in implementation of the toString() method
		// this will keep failing
		return new SigmaDelegateException("Unable to invoke delegate method: '"
				+ delegate.toString() + "' on '" + object + "'", e);
	}

	public SigmaDelegateException handleInvocationTargetException(
			Method delegate, Object object, InvocationTargetException e) {
		e.getTargetException().printStackTrace();

		// FIXME: if the problem is in implementation of the toString() method
		// this will keep failing
		return new SigmaDelegateException("Unable to invoke delegate method: '"
				+ delegate.toString() + "' on '" + object + "': "
				+ e.getTargetException(), e.getTargetException());
	}

	public ValidationResult toSigmaValidationResult(Object status,
			ISigmaValidationDelegate delegate, EObject eObject) {
		boolean result = false;
		String message = null;

		if (status instanceof ValidationResult) {
			return (ValidationResult) status;
		} else if (status instanceof Boolean) {
			// booleans
			result = ((Boolean) status).booleanValue();
		} else if ((status instanceof String)
				|| String.class.isAssignableFrom(delegate
						.getDelegateReturnType())) {
			// string
			result = status == null ? true : false;
			message = (String) status;
		}

		if (!result && message == null) {
			// FIXME: if the problem is in implementation of the toString()
			// method this will keep failing
			message = Utils.bind(
					Messages.Sigma_GenericConstraintViolatedNoMessage,
					delegate.getConstraint(), eObject.toString());
		}

		if (result) {
			return ValidationResult.ok();
		} else {
			return ValidationResult.error(message);
		}
	}

	public String getDelegateConstraintKey() {
		return DELEGATE_CONSTRAINT_KEY;
	}

	public SigmaInvocationDelegateFactory getInvocationDelegateFactory() {
		if (invocationFactory == null) {
			invocationFactory = new SigmaInvocationDelegateFactory();
		}
		return invocationFactory;
	}

	public SigmaSettingDelegateFactory getSettingDelegateFactory() {
		if (settingFactory == null) {
			settingFactory = new SigmaSettingDelegateFactory();
		}

		return settingFactory;
	}

	public SigmaValidationDelegateFactory getValidationDelegateFactory() {
		if (validationFactory == null) {
			validationFactory = new SigmaValidationDelegateFactory();
		}

		return validationFactory;
	}

	public static <T> EList<T> asEList(Collection<T> input) {
		return delegatingEList(ImmutableList.<T> copyOf(input));
	}

	public static <T> EList<T> asEList(Iterable<T> input) {
		return delegatingEList(ImmutableList.<T> copyOf(input));
	}

	public static <T> EList<T> asEList(Iterator<T> input) {
		return delegatingEList(ImmutableList.<T> copyOf(input));
	}
}

File: fr.unice.i3s.sigma/src/fr/unice/i3s/sigma/delegates/SigmaDelegateDomain.scala
package fr.unice.i3s.sigma.delegates

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import scala.collection.JavaConversions._
import org.eclipse.emf.ecore.EClass
import org.eclipse.emf.ecore.EClassifier
import org.eclipse.emf.ecore.EDataType
import org.eclipse.emf.ecore.EGenericType
import org.eclipse.emf.ecore.EOperation
import org.eclipse.emf.ecore.EStructuralFeature
import org.eclipse.emf.ecore.ETypedElement
import org.eclipse.emf.ecore.EValidator
import org.eclipse.emf.common.util.EList
import org.eclipse.emf.ecore.EObject
import java.lang.reflect.Method
import scala.beans.BeanProperty
import fr.unice.i3s.sigma.support.EMFScalaSupport
import fr.unice.i3s.sigma.util.DelegatingEList
import scala.collection.Seq
import java.lang.reflect.InvocationTargetException
import fr.unice.i3s.sigma.support.ecore.EcorePackageScalaSupport
import org.apache.commons.lang3.reflect.TypeUtils
import scala.collection.mutable.Buffer
import fr.unice.i3s.sigma.validation.ValidationResult
import fr.unice.i3s.sigma.validation.Passed
import fr.unice.i3s.sigma.validation.Error

final object SigmaDelegateDomain {

  private val SIGMA_DELEGATE_URI = "http://www.i3s.unice.fr/Sigma"
  private val DELEGATE_ANNOTATION_KEY = "delegate"
  private val DELEGATE_CLASS_SUFFIX = "Delegate"

  private val TYPE_MAPPING = Map("boolean" -> "Boolean")
  private val COLLECTION_TYPE = classOf[Buffer[_]]

  /**
   * Registers globally Java version of the Sigma delegates for the
   * {@link SigmaConstants#SIGMA_DELEGATE_URI} URI. This includes:
   * <ul>
   * <li>Invocation delegate {@link SigmaInvocationDelegate}
   * <li>Setting delegate {@link SettingDelegate}
   * <li>Validation delegate {@link ValidationDelegate}
   * </ul>
   *
   * In order to fully leverage the extended return type of the Sigma
   * validation delegate a {@link SigmaEObjectValidation} must be registered
   * as the package validator. Normally, this is not the case and it has to be
   * done explicitly by changing the generate {@code <PackageName>Validator}
   * super class to be {@link SigmaEObjectValidator}.
   */
  def installGlobally() {
    EOperation.Internal.InvocationDelegate.Factory.Registry.INSTANCE +=
      (instance.getURI -> new SigmaInvocationDelegateFactory);

    EStructuralFeature.Internal.SettingDelegate.Factory.Registry.INSTANCE +=
      (instance.getURI -> new SigmaSettingDelegateFactory);

    EValidator.ValidationDelegate.Registry.INSTANCE +=
      (instance.getURI -> new SigmaValidationDelegateFactory);
  }

  val instance = new SigmaDelegateDomain

}

final class SigmaDelegateDomain extends EMFScalaSupport with EcorePackageScalaSupport {
  import SigmaDelegateDomain._

  def getURI: String = SIGMA_DELEGATE_URI

  def loadDelegateClass(className: String): Option[Class[_]] = {
    val clazz = try {
      Class.forName(className)
    } catch {
      case _: Throwable ⇒
        try {
          Class.forName(className + "$");
        } catch {
          case _: Throwable ⇒ null
        }
    }
    Option(clazz)
  }

  def delegateAnnotationKey: String = DELEGATE_ANNOTATION_KEY

  def classDelegateName(clazz: EClass) = clazz.name + DELEGATE_CLASS_SUFFIX

  def classifierName(c: EClassifier) = {
    c match {
      case t: EDataType ⇒ TYPE_MAPPING.get(t.getInstanceClassName).getOrElse(t.getInstanceClass.getSimpleName)
      case t: EClass ⇒ Option(c.getName).orElse(Option(c.getInstanceTypeName)).getOrElse("")
    }
  }

  def classifierTypeName(c: EClassifier) = {
    c.getName + (c.getETypeParameters match {
      case p if p.isEmpty ⇒ ""
      case p ⇒ p.map(_.getName).mkString("[", ",", "]")
    })
  }

  def classifierTypeParameters(c: EClassifier) = {
    c.getETypeParameters match {
      case p if p.isEmpty ⇒ ""
      case p ⇒ p.map(t ⇒
        t.getName + " <: " + t.getEBounds.map(genericType(_)).mkString("&")).mkString("[", ",", "]")
    }
  }

  def elementTypeName(c: ETypedElement) = {
    var typeName = genericType(c.getEGenericType)

    typeName = if (c.isMany) COLLECTION_TYPE.getSimpleName + "[" + typeName + "]" else typeName
    typeName = if (!c.isRequired && !c.isMany) "Option[" + typeName + "]" else typeName

    typeName
  }

  def genericType(c: EGenericType): String = {
    // taken from EGenericTypeImpl.toString(StringBuilder)

    val r = new StringBuilder
    if (c.getEClassifier != null) {
      r ++= classifierName(c.getEClassifier)

      if (!c.getETypeArguments.isEmpty()) {
        r ++= c.getETypeArguments.map(genericType(_)).mkString("[", ", ", "]")
      }

    } else if (c.getETypeParameter != null) {
      val name = Option(c.getETypeParameter.getName()).getOrElse("")
      r ++= name
    } else {
      r += '_'
      if (c.getELowerBound != null) {
        r ++= " >: "
        r ++= genericType(c.getELowerBound)
      } else {
        if (c.getEUpperBound != null) {
          r ++= " <: "
          r ++= genericType(c.getEUpperBound)
        }
      }
    }

    r.toString
  }

  def getCollectionType() = {
    COLLECTION_TYPE
  }

  def processResult(eType: ETypedElement, value: Any) = {
    value match {
      case o: Seq[_] ⇒ new DelegatingEList(o.toBuffer)
      case o: Option[_] ⇒ o.getOrElse(null)
      case _ ⇒ value
    }
  }

  def safeDynamicInvoke[T](delegateMethod: Method, target: Object)(block: ⇒ T): T = {
    try block
    catch {
      case e: IllegalArgumentException ⇒
        throw handleIllegalArgumentException(delegateMethod, target, e)
      case e: IllegalAccessException ⇒
        throw handleIllegalAccessException(delegateMethod, target, e)
      case e: InvocationTargetException ⇒
        throw handleInvocationTargetException(delegateMethod, target, e)
    }
  }

  protected def handleIllegalArgumentException(delegateMethod: Method, target: Object, e: IllegalArgumentException) = {
    // FIXME: add logging instead of this
    e.printStackTrace();

    // FIXME: if the problem is in implementation of the toString() method
    // this will keep failing
    new SigmaDelegateException(s"Unable to invoke delegate method: `$delegateMethod` on `$target`", e)
  }

  protected def handleIllegalAccessException(delegateMethod: Method, target: Object, e: IllegalAccessException) = {
    // FIXME: add logging instead of this
    e.printStackTrace();

    // FIXME: if the problem is in implementation of the toString() method
    // this will keep failing
    new SigmaDelegateException(s"Unable to invoke delegate method: `$delegateMethod` on `$target`", e)
  }

  protected def handleInvocationTargetException(delegateMethod: Method, target: Object, e: InvocationTargetException) = {
    // FIXME: add logging instead of this
    e.printStackTrace();

    // FIXME: if the problem is in implementation of the toString() method
    // this will keep failing
    new SigmaDelegateException(s"Unable to invoke delegate method: `$delegateMethod` on `$target`: ${e.getTargetException}", e.getTargetException)
  }

  def compatibleType(classifier: EClassifier, sType: Type): Boolean = {
    // FIXME: get rid of TypeUtils
    // FIXME: this only work for simple cases without bounds and arrays
    TypeUtils.isAssignable(sType, classifier.instanceClass)
  }

  def compatibleType(eType: ETypedElement, sType: Type): Boolean = {
    if (eType.required) {
      sType match {
        case c: ParameterizedType if classOf[Option[_]].isAssignableFrom(c.getRawType.asInstanceOf[Class[_]]) ⇒
          compatibleType(eType.getEType, c.getActualTypeArguments()(0))
        case _ ⇒ compatibleType(eType.eType, sType)
      }
    } else if (eType.many) {
      sType match {
        case pType: ParameterizedType ⇒
          getCollectionType.isAssignableFrom(pType.getRawType.asInstanceOf[Class[_]]) &&
            compatibleType(eType.eType, pType.getActualTypeArguments()(0))
        case _ ⇒ false
      }
    } else {
      compatibleType(eType.eType, sType)
    }
  }

  def toValidationResult(status: Any, target: EObject, delegate: SigmaValidationDelegate): ValidationResult =
    status match {
      case o: Boolean if o == true ⇒ Passed
      case o: Boolean if o == false ⇒ Error(s"The `${delegate.constraint}` constraint is violated on `$target`")
      case o: String if classOf[String].isAssignableFrom(delegate.delegateMethod.getReturnType) ⇒
        if (o == null) Passed
        else Error(o)
      case o: ValidationResult ⇒ o
      case _ ⇒ throw new RuntimeException(s"Unexpected return type from `$delegate` while validating `$target`")
    }

  //  private val invocationFactory = new JSigmaInvocationDelegateFactory
  //  def getInvocationDelegateFactory = invocationFactory;

  //  private val settingFactory = new SigmaScalaSettingDelegateFactory
  //  def getSettingDelegateFactory = settingFactory;
  //
  //  private val validationFactory = new SigmaScalaValidationDelegateFactory
  //  def getValidationDelegateFactory = validationFactory;

}
